### ISSUE TITLE
sentiment_analysis.py

### ISSUE THREAD

**Issue by EbrahimHafiz** [EbrahimHafiz/Ibrahim_Hafez_Projects#1]
# Sentiment Analysis Project using Python

# Idea:
# Build a model to classify the sentiment of textual reviews (positive / negative) using a labeled dataset.

# Required Tools:
# - Python
# - Libraries: pandas, numpy, scikit-learn, matplotlib, seaborn, nltk

# --------- 1. Import Libraries ---------
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
import re

nltk.download('stopwords')

# --------- 2. Load Data ---------
# You can download the data from https://ai.stanford.edu/~amaas/data/sentiment/ or use a smaller dataset:
data = pd.read_csv("https://raw.githubusercontent.com/dD2405/Twitter_Sentiment_Analysis/master/train.csv")
data = data[['tweet', 'label']]
data.columns = ['text', 'sentiment']
data['sentiment'] = data['sentiment'].map({0: 'negative', 1: 'positive'})

# --------- 3. Text Preprocessing ---------
stop_words = set(stopwords.words('english'))
stemmer = PorterStemmer()

def preprocess(text):
    text = re.sub(r"[^a-zA-Z]", " ", text)
    text = text.lower().split()
    text = [stemmer.stem(word) for word in text if word not in stop_words]
    return " ".join(text)

data['clean_text'] = data['text'].apply(preprocess)

# --------- 4. Split Data ---------
X = data['clean_text']
y = data['sentiment']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# --------- 5. Convert Text to TF-IDF ---------
vectorizer = TfidfVectorizer(max_features=5000)
X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)

# --------- 6. Train Model ---------
model = MultinomialNB()
model.fit(X_train_tfidf, y_train)

# --------- 7. Evaluation ---------
y_pred = model.predict(X_test_tfidf)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

# --------- 8. Confusion Matrix ---------
cm = confusion_matrix(y_test, y_pred)
sns.heatmap(cm, annot=True, fmt="d", cmap="Blues", xticklabels=['Negative','Positive'], yticklabels=['Negative','Positive'])
plt.xlabel("Predicted")
plt.ylabel("True")
plt.title("Confusion Matrix")
plt.show()

# --------- 9. Test the Model on New Sentence ---------
example = ["I really love this movie, it was amazing!"]
example_clean = [preprocess(x) for x in example]
example_vectorized = vectorizer.transform(example_clean)
prediction = model.predict(example_vectorized)
print("\nExample Prediction:", prediction[0])
